File: MvpDemoA/app/src/main/java/com/example/tonjies/mvpdemoa/module/info/presenter/InfoPresenter.java
package com.example.tonjies.mvpdemoa.module.info.presenter;


import android.content.Context;
import android.util.Log;

import com.example.tonjies.mvpdemoa.module.info.bean.Student;
import com.example.tonjies.mvpdemoa.module.info.view.InfoView;
import com.example.tonjies.mvpdemoa.net.Api;
import com.example.tonjies.mvpdemoa.net.RetrofitHelper;

import io.reactivex.Observer;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import static com.example.tonjies.mvpdemoa.net.RetrofitHelper.getInstance;

public class InfoPresenter {

    //View对象
    private InfoView infoView;

    //Api
    private Api api;

    //绑定View对象
    public void attachView(InfoView view) {
        this.infoView = view;
        //实例化api类
        api = RetrofitHelper.getInstance().getServer();
    }

    //销毁View对象
    public void detachView() {
        infoView = null;
    }

    public void getData() {
        //在此处进行网络请求，并将请求结果返回给Activity，这里假设我们已经请求成功，想要返回学生数据
//        infoView.show("学生姓名：tonjies，学生年龄：20！");
        api.getData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Student>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(Student student) {
                        infoView.show("学生姓名:" + student.getName() + "学生年龄：" + student.getAge());
                    }

                    @Override
                    public void onError(Throwable e) {
                        infoView.show("获取失败"+e.getMessage());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: KMvpDemo1/app/src/main/java/com/example/tonjies/mvp/module/info/presenter/InfoPresenter.kt
package com.example.tonjies.mvp.module.info.presenter

import android.content.Context
import android.util.Log
import com.example.tonjies.mvp.module.info.bean.Student
import com.example.tonjies.mvp.module.info.view.InfoView
import com.example.tonjies.mvp.net.Api
import com.example.tonjies.weatherx.net.RetrofitFactory
import rx.Observable
import rx.Observer
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * 获取学生信息P层类
 */
class InfoPresenter {

    //View接口
    private var infoView: InfoView? = null

    //声明Api接口api
    private var api: Api? = null

    constructor() {
        //初始化RetrofitFactory
        api = RetrofitFactory.instance.create(Api::class.java);
    }


    //绑定View
    fun attachView(view: InfoView) {
        this.infoView = view
    }

    //销毁View对象
    fun detechView() {
        infoView = null
    }

    fun getData() {
        //在此处进行网络请求，并将请求结果返回给Activity，这里假设我们已经请求成功，想要返回学生数据
//        infoView?.show("数据请求成功")

        //进行网络请求
        api!!.getData()
                .observeOn(AndroidSchedulers.mainThread()) //
                .subscribeOn(Schedulers.io())
                .subscribe(object : Observer<Student> {
                    override fun onError(e: Throwable) {
                        Log.d("tonjies", e.message)
                    }

                    override fun onNext(t: Student) {
                        var studentName = t.name //学生姓名
                        var studentAge = t.age //学生年龄
                        infoView!!.show("学生的姓名是:" + studentName + "学生的年龄是:" + studentAge)
                    }

                    override fun onCompleted() {
                    }
                })
    }
}